### ISSUE TITLE
[BUG] Some nodes going down brings down the whole Harvester Cluster

### ISSUE THREAD

**Issue by dipakparmar** [harvester/harvester#5424]
**Describe the bug**

We have observed a critical issue within our 6-node Harvester cluster. Specifically, the last two nodes added to the cluster, when they go down, cause the entire cluster to become unresponsive. This issue prevents access to the kubectl API, and the Harvester UI dashboard becomes unresponsive. Although the login page occasionally loads, attempts to log in result in a redirect back to the login page. Notably, when these nodes are brought back online, we've observed additional nodes going down unexpectedly, despite no changes being made to them. These nodes serve both management and compute roles.

**To Reproduce**
Steps to reproduce the behavior seem to be tied to shutting down nodes randomly. Our cluster setup includes five management nodes. We attempted to demote two of these nodes via lables, intending to repurpose them as worker nodes. However, despite removing the management roles, etcd continues to recognize them as management nodes. This issue has persisted, and we've opened another ticket regarding this matter https://github.com/harvester/harvester/issues/5421. Our cluster configuration is standard, with minimal workloads currently running, including  two virtual machines . The network setup is also standard, without any complex configurations.

**Expected behavior**

Given our cluster has six nodes, we expect the cluster to remain operational even if one or two nodes go down, especially since the design should support functioning with three management nodes. The unresponsiveness and cascading failure of additional nodes is unexpected and severely impacts our operations. Our nodes are equipped with ample disk capacity and are configured with Longhorn for storage, anticipating resilience and redundancy with at least three replicas per data.

**Support bundle**

We have generated support logs from several servers but have not identified any clear indicators related to this issue. We will provide these logs for further analysis privately via email.

**Environment**
 - Harvester ISO version: 1.2.1
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): Baremetal with Dell PowerEdge R620 and R730 servers.

**Additional context**
This issue represents a significant disruption to our cluster's functionality, with a straightforward setup seemingly contributing to an unexpected and severe failure mode. Any insights or guidance on mitigating these issues would be greatly appreciated.

**Comment by dipakparmar** [comment]
Similar https://github.com/harvester/harvester/issues/2203

cc: @ideaman007

**Comment by bk201** [comment]
The maximum count of the management nodes (control plane node with etcd service) in a cluster is 3, is it possible the 2 down nodes are in this role? You can check with `kubectl get nodes`. The control plane requires at least two alive management nodes to function.

**Comment by dipakparmar** [comment]
> The maximum count of the management nodes (control plane node with etcd service) in a cluster is 3, is it possible the 2 down nodes are in this role? You can check with `kubectl get nodes`. The control plane requires at least two alive management nodes to function.

Thanks @bk201 

We have 6 nodes and unfortunately all them have control plane and etcd service so all are management technically. and in this scenario, as we have already have 4 management nodes cluster should be still functioning.

![Shot 2024-03-20 at 21 52 50@2x](https://github.com/harvester/harvester/assets/24366206/3c4b7d56-b0fd-44bd-a45f-7a831bd38c93)

FYI, 5,6 are in Maintenance mode for the testing and to shutdown this scenarios as much safely but that won't be in DR.

**Comment by ibrokethecloud** [comment]
is it possible to get a support bundle? Was anything done to manually promote the nodes, the harvester controller will not try and promote more nodes if 3 are found to have been promoted.

**Comment by dipakparmar** [comment]
> is it possible to get a support bundle? Was anything done to manually promote the nodes, the harvester controller will not try and promote more nodes if 3 are found to have been promoted.

@ibrokethecloud yes, i can provide the support logs via email. Let me if know if you don't find it in harvester-support-bundle@suse.com, I can share it via Slack. Thanks  

No not before the issue, but after the issue, i tried with Kubernetes node role labels.

**Comment by ibrokethecloud** [comment]
we would appreciate a supportbundle, the supportconfig does not provide us any k8s object info. https://docs.harvesterhci.io/v1.3/troubleshooting/harvester#generate-a-support-bundle

**Comment by ibrokethecloud** [comment]
alternative if you could please check how many etcd members you have in the cluster. We do not recommend manually labelling nodes to trigger promotion.

**Comment by dipakparmar** [comment]
> we would appreciate a supportbundle, the supportconfig does not provide us any k8s object info. [docs.harvesterhci.io/v1.3/troubleshooting/harvester#generate-a-support-bundle](https://docs.harvesterhci.io/v1.3/troubleshooting/harvester#generate-a-support-bundle)

Yes, sent support bundle.

**Comment by dipakparmar** [comment]
@ibrokethecloud are you looking for the overall list of the nodes with etcd or output from the etcd cli ?
```etcdctl member list```

**Comment by ibrokethecloud** [comment]
yes. I am unable to access the support bundle from the shared folder. It might be better to upload it here or dm it on rancher user slack

**Comment by dipakparmar** [comment]
@ibrokethecloud sent you via Slack DM.

**Comment by ibrokethecloud** [comment]
The support bundle indicates that you only have 3 etcd members:

```
etcd-hci-04                                              1/1     Running   3 (5d22h ago)      102d    10.80.2.14    hci-04   <none>           <none>
etcd-hci-05                                              1/1     Running   5 (5d20h ago)      102d    10.80.2.15    hci-05   <none>           <none>
etcd-hci-06                                              1/1     Running   5 (5d18h ago)      102d    10.80.2.16    hci-06   <none>           <none>
```

As a result shutting down hci-05 and hci-06 together will cause the outage. 

Harvester will not auto-promote node to being an etcd member until the node exists in the apiserver. 

Ideally the API should have displayed a warning when you tried to place a second etcd member node into maintainence after another etcd member was already put in maintenance.

**Comment by dipakparmar** [comment]
> The support bundle indicates that you only have 3 etcd members:
> 
> ```
> etcd-hci-04                                              1/1     Running   3 (5d22h ago)      102d    10.80.2.14    hci-04   <none>           <none>
> etcd-hci-05                                              1/1     Running   5 (5d20h ago)      102d    10.80.2.15    hci-05   <none>           <none>
> etcd-hci-06                                              1/1     Running   5 (5d18h ago)      102d    10.80.2.16    hci-06   <none>           <none>
> ```
> 
> As a result shutting down hci-05 and hci-06 together will cause the outage.

@ibrokethecloud this make sense to me. 

> Harvester will not auto-promote node to being an etcd member until the node exists in the apiserver.

However, putting the existing management node in maintenance shouldn't the other node be promoted to the management node? Or I'm missing something here? 

> Ideally the API should have displayed a warning when you tried to place a second etcd member node into maintainence after another etcd member was already put in maintenance.

**Comment by ibrokethecloud** [comment]
the promotion logic checks to see if atleast 3 nodes exist which are controlplane nodes. If not, then another one is picked for promotion to controlplane. 

placing a node in maintenance mode does not delete the node from the cluster state(node object still exists). Maintenance is meant to be a temporary outage, and promotion logic will meet the check which successfully finds 3 nodes promoted to controlplane so no further action is needed. A new node will be promoted if you delete node 05 or 06, which which case node will be gone from the apiserver and new promotion will be triggered.

**Comment by dipakparmar** [comment]
> the promotion logic checks to see if atleast 3 nodes exist which are controlplane nodes. If not, then another one is picked for promotion to controlplane.
> 
> placing a node in maintenance mode does not delete the node from the cluster state(node object still exists). Maintenance is meant to be a temporary outage, and promotion logic will meet the check which successfully finds 3 nodes promoted to controlplane so no further action is needed. A new node will be promoted if you delete node 05 or 06, which which case node will be gone from the apiserver and a new promotion will be triggered.

Thanks for clarifying this, however, In DR you can loose two or three nodes and need a way to recover and in that case we need manually promote demote. 

Is there a way to promote and demote nodes or taint them? as we do not want 4,5,6 to be management nodes?

**Comment by pcoueffin** [comment]
I'm seeing the same issue here I think.  It looks like maybe hacking the yaml configs for kubernetes directly might be a temporary fix.  I was looking for other suggestions and stumbled across the suggestion in this thread:

https://github.com/rancher/rancher/issues/22319

It would be nice to have a sanctioned way to promote nodes on a temporary basis in order to keep the cluster stable while we fix the underlying problem though.

**Comment by dipakparmar** [comment]
> > the promotion logic checks to see if atleast 3 nodes exist which are controlplane nodes. If not, then another one is picked for promotion to controlplane.
> > placing a node in maintenance mode does not delete the node from the cluster state(node object still exists). Maintenance is meant to be a temporary outage, and promotion logic will meet the check which successfully finds 3 nodes promoted to controlplane so no further action is needed. A new node will be promoted if you delete node 05 or 06, which which case node will be gone from the apiserver and a new promotion will be triggered.
> 
> Thanks for clarifying this, however, In DR you can loose two or three nodes and need a way to recover and in that case we need manually promote demote.
> 
> Is there a way to promote and demote nodes or taint them? as we do not want 4,5,6 to be management nodes?

@ibrokethecloud checkin back on this, any suggestion ?

**Comment by notsrch** [comment]
This is a huge issue for us.   We lost a very large Harvester cluster, double hardware failure a couple days apart and then the cluster was unrecoverable.   The functionality to automatically avoid this should be a base feature set.

**Comment by w13915984028** [comment]
The label `disaster-recovery` (DR) is added.

Both https://github.com/harvester/harvester/issues/5421 and https://github.com/harvester/harvester/issues/5424 will be taken into account for further enhancements:
(1) Allow user to set e.g. 5 management nodes; the node-promoter can work for more nodes promotion. Also allow to set mode like `auto-promotion` and `manual-promotion`.
(2) Change the node-role on the fly.
(3) Different number of management role and etcd role, e.g. 3 mgmt & 5 etcd, or 5 mgmt & 3 etcd.

TBD...

**Comment by w13915984028** [comment]
@dipakparmar 
>Our cluster setup includes five management nodes. We attempted to demote two of these nodes via lables, intending to repurpose them as worker nodes. However, despite removing the management roles, etcd continues to recognize them as management nodes.

Could you add more details, thanks.

(1) How did you set the additional 2 nodes to be management nodes, by default Harvester sets 3 of them for this role;

(2) `We attempted to demote two of these nodes via lables`, which detailed leables were operated and were those nodes rebooted after changing the labels? did you check the `rke2-server` and `rke2-agent` service on those 2 nodes, any abnormal logs?
